app/master/build_request_handler.py

```diff
@@ -101,6 +101,11 @@ def _prepare_build_async(self, build, project_lock):
                                    log_msg='Build preparation loop is handling request for build {build_id}.')
             try:
                 build.prepare(self._subjob_calculator)
+                # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                if len(build.all_subjobs()) == 0:
+                    self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                    build.mark_started()
```

Comment:
If this line isn't here, the build_fsm raises an exception when the perform_async_postbuild_tasks() tries to progress the FSM to the completed state.

We have two options here:

- either modify the FSM rules to allow the build to go from PREPARED -> FINISHED.
- or do what I did here, where we mark the build as started immediately before performing the asynchronous postbuild tasks.